Sources/BrowserServicesKit/UserScript/ContentScopeUserScript.swift

```diff
@@ -0,0 +1,68 @@
+//
+//  ContentScopeUserScript.swift
+//
+//  Copyright © 2021 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import WebKit
+import Combine
+
+public final class ContentScopePreferences: Encodable {
+    public let globalPrivacyControlValue: Bool
+    public let debug: Bool = false
+    public let sessionKey: String
+
+    public init(gpcEnabled: Bool, sessionKey: String) {
+        self.globalPrivacyControlValue = gpcEnabled
+        self.sessionKey = sessionKey
+    }
+}
+
+public final class ContentScopeUserScript: NSObject, UserScript {
+    public let messageNames: [String] = []
+
+    public init(_ privacyConfigManager: PrivacyConfigurationManager, preferences: ContentScopePreferences) {
+        source = ContentScopeUserScript.generateSource(privacyConfigManager, preferences: preferences)
+    }
+
+    public static func generateSource(_ privacyConfigurationManager: PrivacyConfigurationManager, preferences: ContentScopePreferences) -> String {
+        let privacyConfigJson = String(data: privacyConfigurationManager.currentConfig, encoding: .utf8)
+
+        guard let json = try? JSONEncoder().encode(privacyConfigurationManager.privacyConfig.userUnprotectedDomains),
+              let jsonString = String(data: json, encoding: .utf8),
+              let jsonPreferences = try? JSONEncoder().encode(preferences),
+              let jsonPreferencesString = String(data: jsonPreferences, encoding: .utf8)
+              else {
+            return ""
+        }
+        
+        return loadJS("contentScope", from: Bundle.module, withReplacements: [
+            "$CONTENT_SCOPE$": privacyConfigJson!,
+            "$USER_UNPROTECTED_DOMAINS$": jsonString,
+            "$USER_PREFERENCES$": jsonPreferencesString,
```

Comment:
@jonathanKingston one thing I'm not 100% clear on here, is how each platform will pass platform-specific configuration to the content scripts?

For example, if we want the future navigator interface script to have access to `{ platform: { name: "ios" }` on iOS and `{ platform: { name: "macos" }`, where would this design allow that?

Reply:
Either:

- If we need it remotely configurable through `privacyConfigJson` which is the remote config raw JSON file and we'd parse that out in the relevant feature file.
- Or we can pass it through `ContentScopePreferences` for constants like you're referring to.

The platform itself is responsible for the relevant `inject/platform.js` script and that will likely be different in how this data gets wired into the calling API of `contentScopeFeatures.init()` (eventually this will be typescript checked so it's clear what this API is)

Reply:
https://github.com/duckduckgo/content-scope-scripts/blob/main/inject/apple.js#L53 is where these call sites go in and this combines the data in the format expected by `contentScopeFeatures.init(processedConfig)`